From 47a7154033f8785579f109f1f48e9457b4e487f0 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Fri, 13 Dec 2024 11:42:42 +0100
Subject: [PATCH 1301/1695] drivers: stm32_uart: handle clock disable during PM

During PM, console clock could be disabled. So we cannot access to
register during suspend to save them, and we cannot print traces.

First, do not save register values during suspend but at probe instead.
And then check is the clock is enabled before printing a trace.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c65f9bedb56bc34d83c53198191faa2286468f36 drivers: stm32_uart: handle clock disable during PM]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/424706
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/430773
Change-Id: Ia20b1d873007d45d9df47b6182bdd9fbdb0b3624
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/436509
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_uart.c         | 39 +++++++++++++++++++++++++++++++
 core/include/drivers/stm32_uart.h |  1 +
 2 files changed, 40 insertions(+)

diff --git a/core/drivers/stm32_uart.c b/core/drivers/stm32_uart.c
index 81d5da29c..a7bfc3627 100644
--- a/core/drivers/stm32_uart.c
+++ b/core/drivers/stm32_uart.c
@@ -66,6 +66,19 @@
 #define UART_REG_BRR_M_SHIFT		4
 #define UART_REG_BRR_MASK		GENMASK_32(15, 0)
 
+static bool loc_chip_clk_is_enabled(struct serial_chip *chip)
+{
+	struct stm32_uart_pdata *pd = container_of(chip,
+						   struct stm32_uart_pdata,
+						   chip);
+
+	if (pd->clock && pd->clock->ops->is_enabled)
+		return pd->clock->ops->is_enabled(pd->clock);
+
+	/* For early console, assume the clock is enable */
+	return true;
+}
+
 static vaddr_t loc_chip_to_base(struct serial_chip *chip)
 {
 	struct stm32_uart_pdata *pd = NULL;
@@ -80,6 +93,9 @@ static void loc_flush(struct serial_chip *chip)
 	vaddr_t base = loc_chip_to_base(chip);
 	uint64_t timeout = timeout_init_us(FLUSH_TIMEOUT_US);
 
+	if (!loc_chip_clk_is_enabled(chip))
+		return;
+
 	while (!(io_read32(base + UART_REG_ISR) & USART_ISR_TXFE))
 		if (timeout_elapsed(timeout))
 			return;
@@ -90,6 +106,9 @@ static void loc_putc(struct serial_chip *chip, int ch)
 	vaddr_t base = loc_chip_to_base(chip);
 	uint64_t timeout = timeout_init_us(PUTC_TIMEOUT_US);
 
+	if (!loc_chip_clk_is_enabled(chip))
+		return;
+
 	while (!(io_read32(base + UART_REG_ISR) & USART_ISR_TXE_TXFNF))
 		if (timeout_elapsed(timeout))
 			return;
@@ -260,6 +279,15 @@ static TEE_Result stm32_uart_pm(enum pm_op op, uint32_t pm_hint,
 	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT))
 		return TEE_SUCCESS;
 
+	if (pd->is_console && !pd->clock->ops->is_enabled(pd->clock)) {
+		/*
+		 * Maybe the console clock is disabled by non-secure
+		 * or disabled after wake-up from standby,
+		 * in this case do nothing.
+		 */
+		return TEE_SUCCESS;
+	}
+
 	switch (op) {
 	case PM_OP_SUSPEND:
 		pd->brr = io_read32(uart_base + UART_REG_BRR);
@@ -309,10 +337,21 @@ static TEE_Result stm32_uart_probe(const void *fdt, int node,
 
 	if (uart_is_for_console((void *)fdt, node, &params) &&
 	    stm32_uart_update_baudrate(pd, params) == TEE_SUCCESS) {
+		vaddr_t uart_base = io_pa_or_va(&pd->base, 1);
+
 		register_serial_console(&pd->chip);
 		IMSG("UART console (%ssecure)", pd->secure ? "" : "non-");
+
+		pd->is_console = true;
+		pd->brr = io_read32(uart_base + UART_REG_BRR);
+		pd->presc = io_read32(uart_base + UART_REG_PRESC);
+		pd->cr1 = io_read32(uart_base + UART_REG_CR1);
+		pd->cr2 = io_read32(uart_base + UART_REG_CR2);
+		pd->cr3 = io_read32(uart_base + UART_REG_CR3);
 	}
 
+	/* This clock ops is required for PM */
+	assert(pd->clock->ops->is_enabled);
 	register_pm_core_service_cb(stm32_uart_pm, pd, "stm32-uart");
 
 out:
diff --git a/core/include/drivers/stm32_uart.h b/core/include/drivers/stm32_uart.h
index de3b0a454..7088975fe 100644
--- a/core/include/drivers/stm32_uart.h
+++ b/core/include/drivers/stm32_uart.h
@@ -17,6 +17,7 @@ struct stm32_uart_pdata {
 	struct io_pa_va base;
 	struct serial_chip chip;
 	bool secure;
+	bool is_console;
 	struct clk *clock;
 	struct pinctrl_state *pinctrl;
 	uint32_t brr;
-- 
2.47.2

